assertsprocessor/request_context.go

```diff
@@ -0,0 +1,122 @@
+package assertsprocessor
+
+import (
+	"go.opentelemetry.io/collector/pdata/ptrace"
+	"go.uber.org/zap"
+	"reflect"
+	"regexp"
+)
+
+type requestConfigCompiled struct {
+	attrName    string
+	spanKind    string
+	regex       *regexp.Regexp
+	replacement string
+}
+
+type requestContextBuilder interface {
+	getRequest(span *ptrace.Span, serviceKey string) string
+}
+
+type requestContextBuilderImpl struct {
+	logger         *zap.Logger
+	requestConfigs map[string][]*requestConfigCompiled
+}
+
+func (rCB *requestContextBuilderImpl) compileRequestContextRegexps(config *Config) error {
+	rCB.logger.Info("Compiling request context regexps")
+	requestConfigs := make(map[string][]*requestConfigCompiled)
+	if config.RequestContextExps != nil {
+		for serviceKey, serviceRequestContextExps := range config.RequestContextExps {
+			configs := make([]*requestConfigCompiled, 0)
+			for _, matcher := range serviceRequestContextExps {
+				compile, err := regexp.Compile(matcher.Regex)
+				if matcher.SpanKind == "" {
+					matcher.SpanKind = "Server"
+				}
+				rCB.logger.Debug("Compiled request context regex",
+					zap.String("Service", serviceKey),
+					zap.String("Span Kind", matcher.SpanKind),
+					zap.String("AttrName", matcher.AttrName),
+					zap.String("Regex", matcher.Regex),
+					zap.String("Replacement", matcher.Replacement),
+				)
+				if err != nil {
+					return err
+				}
+				replacement := matcher.Replacement
+				if replacement == "" {
+					replacement = "$1"
+				}
+				configs = append(configs, &requestConfigCompiled{
+					attrName:    matcher.AttrName,
+					spanKind:    matcher.SpanKind,
+					regex:       compile,
+					replacement: replacement,
+				})
+			}
+			requestConfigs[serviceKey] = configs
+		}
+		rCB.requestConfigs = requestConfigs
+	}
+	rCB.logger.Debug("Compiled request context regexps successfully")
+	return nil
+}
+
+func (rCB *requestContextBuilderImpl) getRequest(span *ptrace.Span, serviceKey string) string {
+	var request string
+	if rCB.requestConfigs[serviceKey] != nil {
```

Comment:
The earlier logic used to go through the expressions for a specific service and if the request context is not built it would then go through the default expressions. With that approach the default expressions of looking for a http.route value or using the first 2 segments can remain under default. With this logic change, we would have to add the default expressions to all service specific overrides